File: core/memory.py
import sqlite3

DB = "memory.db"


def init_db():
    conn = sqlite3.connect(DB)

    conn.execute("""
    CREATE TABLE IF NOT EXISTS messages(
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        uid INTEGER,
        role TEXT,
        text TEXT
    )
    """)

    conn.commit()
    conn.close()


def save(uid, role, text):

    conn = sqlite3.connect(DB)

    conn.execute(
        "INSERT INTO messages(uid,role,text) VALUES(?,?,?)",
        (uid, role, text)
    )

    conn.commit()
    conn.close()


def load(uid, limit=20):

    conn = sqlite3.connect(DB)

    rows = conn.execute(
        "SELECT role,text FROM messages WHERE uid=? ORDER BY id DESC LIMIT ?",
        (uid, limit)
    ).fetchall()

    conn.close()

    rows.reverse()

    history = []

    for r in rows:
        history.append({
            "role": r[0],
            "content": r[1]
        })

    return history

File: src/agent/memory.ts
import { db } from '../core/db';

export function saveMemory(uid: number, key: string, value: string) {
  db.prepare(`INSERT INTO memory (uid, key, value) VALUES (?, ?, ?)
    ON CONFLICT(uid, key) DO UPDATE SET value=excluded.value`).run(uid, key, value);
}

export function getMemories(uid: number): Array<{ key: string; value: string }> {
  return db.prepare('SELECT key, value FROM memory WHERE uid = ? ORDER BY created_at DESC LIMIT 20').all(uid) as any;
}

export function clearMemory(uid: number) {
  db.prepare('DELETE FROM memory WHERE uid = ?').run(uid);
}

export function saveMessage(uid: number, role: 'user' | 'assistant', content: string) {
  db.prepare('INSERT INTO conversations (uid, role, content) VALUES (?, ?, ?)').run(uid, role, content);
  // Keep only last 30 messages
  db.prepare(`DELETE FROM conversations WHERE uid = ? AND id NOT IN (
    SELECT id FROM conversations WHERE uid = ? ORDER BY id DESC LIMIT 30
  )`).run(uid, uid);
}

export function getHistory(uid: number, limit = 10): Array<{ role: string; content: string }> {
  return (db.prepare('SELECT role, content FROM conversations WHERE uid = ? ORDER BY id DESC LIMIT ?').all(uid, limit) as any).reverse();
}

export function clearHistory(uid: number) {
  db.prepare('DELETE FROM conversations WHERE uid = ?').run(uid);
}
